packages/sdk/src/transactions/PSBTBuilder.ts

```diff
@@ -0,0 +1,468 @@
+          this.changeAmount * -1
+        } sats`,
+      );
+    }
+  }
+
+  private getRetrievedUTXOsValue() {
+    return this.utxos.reduce((acc, utxo) => acc + utxo.sats, 0);
+  }
+
+  private getReservedUTXOs() {
+    return this.utxos.map((utxo) =>
+      generateTxUniqueIdentifier(utxo.txid, utxo.n),
+    );
+  }
+
+  private getUTXOAmountToRequestFromChangeAmount() {
+    if (this.changeAmount < 0) {
+      return Math.abs(this.changeAmount);
+    }
+
+    return this.outputAmount - this.getRetrievedUTXOsValue();
+  }
+
+  private async retrieveUTXOs(address?: string, amount?: number) {
+    if (!this.autoAdjustment && !address) {
+      return;
+    }
+
+    const amountToRequest =
+      amount && amount > 0
+        ? amount
+        : this.getUTXOAmountToRequestFromChangeAmount();
+    if (
+      (amount && this.getRetrievedUTXOsValue() >= amount) ||
+      amountToRequest <= 0
+    ) {
+      return;
+    }
+
+    const utxos = await this.datasource.getSpendables({
+      address: address || this.address,
+      value: convertSatoshisToBTC(amountToRequest),
+      filter: this.getReservedUTXOs(),
+    });
+
+    this.noMoreUTXOS = utxos.length === 0;
+
+    this.utxos.push(...utxos);
+  }
+
+  protected async retrieveSelectedUTXOs(address: string, amount: number) {
+    await this.retrieveUTXOs(address, amount);
+    const selectedUTXOs = this.utxos.find((utxo) => utxo.sats >= amount);
+
+    this.utxos = selectedUTXOs ? [selectedUTXOs] : [];
+
+    return this.utxos;
+  }
+
+  private async prepareInputs() {
+    if (!this.autoAdjustment) {
+      return undefined;
+    }
+
+    const promises: Promise<InputType>[] = [];
+
+    this.utxos.forEach((utxo) => {
+      if (
+        this.usedUTXOs.includes(generateTxUniqueIdentifier(utxo.txid, utxo.n))
+      ) {
+        return;
+      }
+
+      this.inputAmount += utxo.sats;
+      const promise = processInput({
+        utxo,
+        pubKey: this.publicKey,
+        network: this.network,
+        datasource: this.datasource,
+      }); // TODO: add sigHashType
+
+      promises.push(promise);
+    });
+
+    const response = await Promise.all(promises);
+
+    this.inputAmount += this.injectableInputs.reduce(
+      (acc, curr) => acc + curr.sats,
+      0,
+    );
+    response.forEach((input) => {
+      const txUniqueIdentifier = generateTxUniqueIdentifier(
+        input.hash,
+        input.index,
+      );
+      if (this.usedUTXOs.includes(txUniqueIdentifier)) {
+        return;
+      }
+      this.usedUTXOs.push(txUniqueIdentifier);
+    });
+
+    this.inputs = this.inputs.concat(response);
+
+    return this.inputs;
+  }
+
+  async prepare() {
+    // calculate output amount
+    this.calculateOutputAmount();
+
+    // fetch UTXOs to spend
+    await this.retrieveUTXOs();
+    await this.prepareInputs();
+
+    await this.calculateChangeAmount();
+
+    this.process();
+    await this.calculateChangeAmount();
+    this.calculateOutputAmount();
+
+    this.process();
```

Comment:
@iamcrazycoder why are we calling `process` twice?

Reply:
There's always a chance that a change output will be added after calculating the network fee. `PSBTBuilder` re-processes everything to ensure it has enough UTXOs to cover for the outputs (including change) + network fee.